### DISCUSSION TITLE
【Feature Request】Server Function

### DISCUSSION THREAD

**Discussion by x1ny** [vercel/next.js#37294] [Ideas]
### Describe the feature you'd like to request

As React is working on server-side components for do some server-side work in the front-end project,I thinking of implementing isomorphic application in nextjs framework.In nextjs, we can write api route and call it in client code.But this method is not easy enough, especially when you have more and more apis,this will require significant maintenance costs.To make the project more structured, we usually put logically related files in the same folder on a page by page basis.So I want to implement a special file type called "Server Function", this js module export a function,when nextjs building, this function will automatically generate an API route.In client code,we can call this function directly, and SWC/Babel will convert it to calling API route.

### Describe the solution you'd like

Specify a module that defualt export a function, this function will automatically generate an API route.If client code call this function, SWC/Babel will convert it to calling API route.

### Describe alternatives you've considered

no

**Comment by icyJoseph** [comment]
There's on way to do this, but it requires `getServerSideProps`, so it is bound to be rather slow, compared to other options anyway.

If you check `context.req.method`, you can receive pretty much any type of request. For example for `GET`, you serve the page and for `POST` you receive form data.

There's no need  to reinvent the wheel, https://next-runtime.meijer.ws/getting-started/1-introduction, does this for you.

**Reply by x1ny** [reply]
This is not what I want, what I need is something like:


```
../../some_page/fetch_some_data.server_function.js

// this is a function that will be run on the server
export default async function(params) {

    ... // some code to fetch data from database

    return data
}
```

```
../../some_page/index.jsx

import fetch_some_data from './fetch_some_data.server_function.js'

// this is a page
export default function() {

   const params = {...}

    return (
       <button
           onClick={() => {
              fetch_some_data(params)
                  .then(data => {
                       console.log(data)
                   })
           }}
       >
           fetch data
       </button>
    )
}
```
When you are coding, It's like a simple function call, so you can easily execute some code on the server, and it is easier to maintain your code.Especially when you use typescript, you will get type hints to improve development efficiency.

At compile time, "fetch_some_data.server_function.js" will be converted to an api route, and the client code will also be converted to look like this:

```
../../some_page/index.jsx

// this is a page
export default function() {

   const params = {...}

    return (
       <button
           onClick={() => {
              fetch('/api/xxxxxx', {
                  method: 'post',
                  headers: {
                      'Accept': 'application/json',
                      'Content-Type': 'application/json'
                  },
                  body: JSON.stringify(params)
              })
                .then(data => {
                  console.log(data)
                })
           }}
       >
           fetch data
       </button>
    )
}
```
